perun/collect/trace/strategy.py

```diff
@@ -28,213 +28,287 @@ class _Status(IntEnum):
-       - the rules are stored as lists of dictionaries with keys 'name', 'sample' and optionally
-         'pair' for rules that can be paired (such as static or dynamic rules)
+     - func, static: contains prepared rules for profiling with unified sampling specification
+       - the rules are stored as dict of dictionaries with keys 'name', 'sample' and optionally
+         'pair' for rules that can be paired (such as static rules)
 
     :param list func: the list of function names that will be traced
     :param list func_sampled: the list of function names with specified sampling
     :param list static: the list of static probes that will be traced
     :param list static_sampled: the list of static probes with specified sampling
-    :param list dynamic: the list of dynamic probes that will be traced
-    :param list dynamic_sampled: the list of dynamic probes with specified sampling
     :param kwargs: additional configuration parameters
 
-    :returns kwargs: the updated dictionary with post processed rules
+    :return kwargs: the updated dictionary with post processed rules
     """
+    # Do some strategy specific actions (e.g. extracting symbols)
+    kwargs['global_sampling'], extracted_func, extracted_static = _strategy_specifics(**kwargs)
 
-    # Remove duplicate rules, merge sampled / non-sampled rule lists and optionally pair the rules
-    kwargs['func'] = _remove_duplicate_probes(_merge_probes_lists(
-        func, func_sampled, kwargs['global_sampling']))
-    kwargs['static'] = []
-    if kwargs['with_static']:
-        kwargs['static'] = _remove_duplicate_probes(_pair_rules(_merge_probes_lists(
-            static, static_sampled, kwargs['global_sampling'])))
-    kwargs['dynamic'] = _remove_duplicate_probes(_pair_rules(_merge_probes_lists(
-        dynamic, dynamic_sampled, kwargs['global_sampling'])))
+    # Create one dictionary of function probes specification
+    kwargs['func'] = _merge_probes(func, func_sampled, extracted_func, kwargs['global_sampling'])
 
-    # Build sampling dictionary
+    # Create one dictionary of static probes specification
+    kwargs['static'] = extracted_static
+    if kwargs['with_static']:
+        # We also need to do some automated pairing of static rules
+        kwargs['static'] = _pair_rules(
+            _merge_probes(static, static_sampled, extracted_static, kwargs['global_sampling']))
     return kwargs
 
 
-def extraction_strategy(func, func_sampled, static, static_sampled, **kwargs):
-    """The userspace, all, u_sampled and a_sampled strategy implementation. No manual
-    configuration of profiled locations is required and is instead automatically extracted
-    from the provided binary according to the specific strategy. However the user can
-    specify additional rules, that have higher priority than the extracted ones.
+def _strategy_specifics(method, binary, with_static, global_sampling, **_):
```

Comment:
I'm not particularly fond of this prototype. I suggest  the following: (1) rename the method parameter to strategy, as it more complies to the module and the function (if there is clash with the module itself then maybe "used_strategy" or something similar), and (2) rename the method to something that is a verb ;-), e.g. _get_strategy_specific or _extract_strategy_specifics.